### PR TITLE
Enable SelectionArea double tap/triple tap gesture support for mobile platforms

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
This change enables double tap / triple tap support in SelectionArea for mobile platforms:
Android / Fuchsia: 
- On native, these platforms allow for double tap / double tap + drag to select word-by-word.
- On web using touch, these platforms only support double tap to select word.
- On web and native using a mouse, these platforms support double click / double click + drag to select word-by-word, and triple click / triple click + drag to select paragraph-by-paragraph.

iOS:
- On native, these platforms allow for double tap / double tap + drag to select word-by-word.
- On web using touch, these platforms do not support double tap/triple tap gestures.
- On web using touch, these platforms allow support double tap + drag gestures.
- On web and native using a mouse, these platforms support double click / double click + drag to select word-by-word, and triple click / triple click + drag to select paragraph-by-paragraph.

Part of: https://github.com/flutter/flutter/issues/129583

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 1]
```diff
@@ -487,18 +469,52 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
...
       case TargetPlatform.fuchsia:
+        if (_lastPointerDeviceKind != null && _lastPointerDeviceKind != PointerDeviceKind.mouse) {
+          // When the pointer device kind is not precise like a mouse, native
+          // Android resets the tap count at 2. For example, this is so the
+          // selection can collapse on the third tap.
+          maxConsecutiveTap = 2;
+        }
+        // From observation, these platforms reset their tap count to 0 when
+        // the number of consecutive taps exceeds the max consecutive tap supported.
+        // For example on Debian Linux with GTK, when going past a triple click,
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:513]: but linux is not in this switch case, it is fuchsia instead.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 2]
```diff
@@ -519,19 +535,41 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
...
-          ..onDragEnd = _handleMouseDragEnd
-          ..onCancel = _clearSelection
-          ..dragStartBehavior = DragStartBehavior.down;
-      },
-    );
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+      case TargetPlatform.iOS:
+        _gestureRecognizers[TapAndHorizontalDragGestureRecognizer] = GestureRecognizerFactoryWithHandlers<TapAndHorizontalDragGestureRecognizer>(
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:542]: Can you add some comment on why we use tap and horizontal drag here? also is fuchsia meant to be here?

File: packages/flutter/test/widgets/selectable_region_test.dart [hunk 3]
```diff
@@ -60,14 +60,14 @@ void main() {
       await tester.pumpAndSettle();
       renderSelectionSpy.events.clear();
 
-      await gesture.moveTo(const Offset(200.0, 100.0));
+      await gesture.moveTo(const Offset(100.0, 200.0));
```
Comment (chunhtai) [inline packages/flutter/test/widgets/selectable_region_test.dart:63]: why this change?

File: packages/flutter/test/widgets/selectable_region_test.dart [hunk 4]
```diff
@@ -596,6 +627,467 @@ void main() {
...
+      final RenderParagraph paragraph3 = tester.renderObject<RenderParagraph>(find.descendant(of: find.text('Fine, thank you.'), matching: find.byType(RichText)));
+      await gesture.moveTo(textOffsetToPosition(paragraph3, 6));
+      expect(paragraph1.selections[0], const TextSelection(baseOffset: 0, extentOffset: 12));
+      expect(paragraph2.selections[0], const TextSelection(baseOffset: 0, extentOffset: 14));
+      expect(paragraph3.selections[0], const TextSelection(baseOffset: 0, extentOffset: 11));
+
+      await gesture.up();
+    },
+      variant: TargetPlatformVariant.mobile(),
+      skip: kIsWeb, // [intended] Double tap / double tap + drag gestures are not fully enabled on the web for mobile platforms.
```
Comment (chunhtai) [inline packages/flutter/test/widgets/selectable_region_test.dart:755]: why are they not enabled? are they going to be enabled in the future?

Comment (Renzo-Olivares) [reply packages/flutter/test/widgets/selectable_region_test.dart:755]: This is the behavior I observed on native platforms (TextView on Android and UITextView on iOS), and while testing on different websites on the web. I'm not against enabling them, what do you think?

I will enable double tap + drag on iOS since that looks like it works on the web, but from my observations a regular double tap does nothing (sometimes zooms in page).

Comment (chunhtai) [reply packages/flutter/test/widgets/selectable_region_test.dart:755]: oh I see, In that case I will probably rephrase this to say something like

```
web does not support double tap + drag gestures.
```

Comment (Renzo-Olivares) [reply packages/flutter/test/widgets/selectable_region_test.dart:755]: Sounds good, will do that.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 2 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/selectable_region.dart]: Within `TextSelectionGestureDetector` we default to grouping fuchsia with Android if the behavior is unknown.

File: packages/flutter/test/widgets/selectable_region_test.dart (hunk 3 above)
Comment (Renzo-Olivares) [reply packages/flutter/test/widgets/selectable_region_test.dart]: This was because I was using `TapAndHorizontalDragGestureRecognizer`, but I have since changed it to `TapAndPan` so this change is reverted.

**Review by chunhtai — APPROVED** [review]
LGTM
